tests/unit/s2n_traffic_context_test.c

```diff
@@ -0,0 +1,431 @@
+
+        /* Preliminary send and receive */
+        EXPECT_OK(s2n_test_send_and_recv(server_conn, client_conn));
+        EXPECT_OK(s2n_test_send_and_recv(client_conn, server_conn));
+
+        uint8_t buffer[S2N_TRAFFIC_CONTEXT_TLS12_SIZE] = { 0 };
+        EXPECT_SUCCESS(s2n_connection_get_traffic_context(client_conn, buffer, sizeof(buffer)));
+
+        /* Initialize new client connection and set traffic context */
+        DEFER_CLEANUP(struct s2n_connection *new_client_conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(new_client_conn);
+        EXPECT_SUCCESS(s2n_connection_set_traffic_context(new_client_conn, buffer, sizeof(buffer)));
+
+        /* Re-initialize IO pipes */
+        EXPECT_OK(s2n_connections_set_io_stuffer_pair(new_client_conn, server_conn, &io_pair));
+
+        /* Client can send and recv as usual */
+        EXPECT_OK(s2n_test_send_and_recv(server_conn, new_client_conn));
+        EXPECT_OK(s2n_test_send_and_recv(new_client_conn, server_conn));
+    };
+
+    /* Self-talk: Server can be serialized and deserialized and continue sending and receiving data in TLS1.2 and TLS1.3 */
+    for (size_t i = 0; i < s2n_array_len(config_array); i++) {
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client_conn);
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config_array[i]));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config_array[i]));
+
+        DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+        EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+        EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, server_conn, &io_pair));
+
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        /* Preliminary send and receive */
+        EXPECT_OK(s2n_test_send_and_recv(server_conn, client_conn));
+        EXPECT_OK(s2n_test_send_and_recv(client_conn, server_conn));
+
+        uint8_t buffer[S2N_TRAFFIC_CONTEXT_TLS12_SIZE] = { 0 };
+        EXPECT_SUCCESS(s2n_connection_get_traffic_context(server_conn, buffer, sizeof(buffer)));
+
+        /* Initialize new server connection and set traffic context */
+        DEFER_CLEANUP(struct s2n_connection *new_server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(new_server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_traffic_context(new_server_conn, buffer, sizeof(buffer)));
+
+        /* Re-initialize IO pipes */
+        EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, new_server_conn, &io_pair));
+
```

Comment:
Does this keep any leftover data in the stuffers? You probably need a fresh io_pair to do these tests right.